packages/gateway-client/src/worker/p2p-client/streams.ts

```diff
@@ -0,0 +1,372 @@
+            if (headLength === 0) {
+                headLength = chunk.readUInt16BE(0) + 2;
+                this.log.trace('headLength', headLength);
+            }
+
+            // add the length of the current chunk to the total length
+            currentLength += chunk.length;
+
+            // if we haven't read the packet json yet, we don't know the
+            // total length of the packet, so we can't know when we're done
+            let packet = null;
+            if (totalLength === 0 && currentLength >= headLength) {
+                packet = decode(Buffer.concat(chunks));
+                totalLength =
+                    ((packet?.json?.bodyLength as number) ?? 0) + headLength;
+                this.log.trace('totalLength', totalLength);
+            }
+
+            // if we've read the packet json and we've got the total length
+            // of the packet, we can resolve the inbox
+            if (currentLength === totalLength) {
+                // ensure we have the packet
+                packet = packet || decode(Buffer.concat(chunks));
+                // reset our local state
+                currentLength = headLength = totalLength = 0;
+                chunks = [];
+                // resolve the inbox
+                this.log.trace('inbox resolved', packet);
+                this._receive(packet as Packet);
+            }
+
+            this.emit('chunk', chunk);
+        }
+
+        this.log.debug('stream stopped receiving data');
+    }
+
+    public send(packet: Packet): void {
+        this.outbox.resolve(packet);
+        this.outbox = new Deferred<Packet>();
+    }
+
+    public receive(): Promise<Packet | null> {
+        return this.inbox.promise;
+    }
+
+    private async _receive(packet: Packet | null): Promise<void> {
+        this.inbox.resolve(packet);
+        this.inbox = new Deferred<Packet | null>();
+    }
+}
+
+export class PooledLobStream extends LobStream {
+    private static pool: Set<PooledLobStream> = new Set();
+
+    static getFromPool(
+        protocol: string,
+        Constructor: typeof PooledLobStream
+    ): PooledLobStream | null {
+        for (const potential of this.pool) {
+            if (!potential.isOpen) {
+                this.pool.delete(potential);
+                continue;
+            }
+
+            if (
+                potential.protocol === protocol &&
+                potential.constructor === Constructor
+            ) {
+                this.pool.delete(potential);
+                return potential;
+            }
+        }
+
+        return null;
+    }
+
+    public release() {
+        PooledLobStream.pool.add(this);
+    }
+}
+
+export class RequestStream extends PooledLobStream {
+    public async request(
+        packet: Packet,
+        onChunk?: Callback
+    ): Promise<Packet | null> {
+        if (onChunk) {
+            this.addListener('chunk', onChunk);
```

Comment:
I chose `events` because I incorrectly remembered EventTarget not being in node, and I wanted streams to be isomorphic for eventual re-use in the proxy -> athena package refactor, but I see I was mistaken, and will change. 

I'm 50/50 on whether I want consumers to have access to `stream.on(` stuff.